### PR TITLE
Adding band structure parser for cp2k version 8.1.

### PR DESCRIPTION
**Pull request by hdsassnick**
Hello,

I would firstly like to thank you for the work and effort developing the AiiDA-plugin for cp2k.
It is a great addition to the AiiDA-library and facilitates my work a lot.

I noticed that for the new version 8.1 of cp2k the output format of the band structure has changed and is incompatible with previous versions.
I would like to contribute an adaption of the advanced-output parser which should work with the most recent versions of cp2k
but still keeps the downwards compatibility to older versions of the program.

I would be very happy if the snippet of additional code can be helpful.
This is my first pull request. Apologies if I hadn't done it completely correctly.

Best wishes,
Holger

### CODE REVIEW

**Comment by yakutovicha** [comment]
Hey @hdsassnick, thanks a lot. 

> I would be very happy if the snippet of additional code can be helpful.

No worries, I think you did a great job.

Before I start reviewing the PR, can you please fix the pre-commit issues? It basically just adds two empty lines, see [here](https://github.com/aiidateam/aiida-cp2k/pull/126/checks?check_run_id=1901111007)

**Comment by hdsassnick** [comment]
Hi, should work now.

**Comment by hdsassnick** [comment]
Hi,

I was just wondering if there was any news concerning the review of the pull request.

**Review by yakutovicha — CHANGES_REQUESTED** [review]
Hey @hdsassnick, sorry for the long wait.

I had a closer look at your implementation and I have a couple of suggestions.  Generally, I think it would be a good idea to reuse the code that is there already. For instance, the parts of the code that don't work with the 8.1 version of CP2K are:

https://github.com/aiidateam/aiida-cp2k/blob/60d518309055f9731360db81ead3b2119a154b0f/aiida_cp2k/utils/parser.py#L301

and

https://github.com/aiidateam/aiida-cp2k/blob/60d518309055f9731360db81ead3b2119a154b0f/aiida_cp2k/utils/parser.py#L312-L317

Herein, I think it would be better to modify the `_parse_bands` function as follows

```python

def _parse_band_before81(band_lines):
    put the old code here

def _parse_band_after81(band_lines):
    put the new code here

def _parse_bands(lines, n_start, cp2k_version):
    if cp2k_version < 8.1:
        pattern = re.compile(".*?Nr.*?Spin.*?K-Point.*?", re.DOTALL)
        parse_band_function = _parse_band_before81
    else:
        pattern = re.compile(".*?Point.*?Spin.*?", re.DOTALL) # didn't try it myself
        parse_band_function = _parse_band_after81

```

I think this approach allows us to write shorter and clearer code.

Let me know if you agree.

**Comment by hdsassnick** [comment]
Hi,
Thank you very much for the suggestions. Indeed I forgot to remove some print statements that I just included for testing the function, thank you for pointing that out.
I think the differences between the two functions is actually a bit larger, e.g. the number of bands is not printed at the head of each k-point in the new cp2k version. 
Thus, the current function for cp2k v8.1 is checking line-by-line if an eigenvalue needs to be added to the list. In contrast to that, in cp2k v7.1 the number of eigenvalues is known beforehand, and this information is then being used to add all eigenvalues in one go.
That is why I have a few doubts that merging the two versions of "_parse_bands" would simplify the code by a large extent. Nevertheless, I will give it a try.

**Comment by hdsassnick** [comment]
Hi @yakutovicha ,

I tried to implement your suggestions. I'm not sure if I did it the way you intended.
Could you please have another look at the implementation?

File: test/test_parser.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
+# -*- coding: utf-8 -*-
+###############################################################################
+# Copyright (c), The AiiDA-CP2K authors.                                      #
+# SPDX-License-Identifier: MIT                                                #
+# AiiDA-CP2K is hosted on GitHub at https://github.com/aiidateam/aiida-cp2k        #
+# For further information on the license, see the LICENSE.txt file.           #
+###############################################################################
+"""Test writing structures to xyz format"""
```
Comment (yakutovicha) [inline test/test_parser.py:8]:
```suggestion
"""Test parsing band structure."""
```

File: test/test_parser.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
+# -*- coding: utf-8 -*-
+###############################################################################
+# Copyright (c), The AiiDA-CP2K authors.                                      #
+# SPDX-License-Identifier: MIT                                                #
+# AiiDA-CP2K is hosted on GitHub at https://github.com/aiidateam/aiida-cp2k        #
+# For further information on the license, see the LICENSE.txt file.           #
+###############################################################################
+"""Test parsing band structure."""
```
Comment (yakutovicha) [inline test/test_parser.py:8]:
```suggestion
"""Test output parser."""
```

File: aiida_cp2k/utils/parser.py [hunk 3]
```diff
@@ -282,8 +285,34 @@ def parse_cp2k_output_advanced(fstring):  # pylint: disable=too-many-locals, too
...
+
+    splitted = lines[line_n].split()
+    spin = int(splitted[3])
+    kpoint = tuple(float(p) for p in splitted[-3:])
+    nlines = int(math.ceil(int(lines[line_n + 1]) / 4))
+    bands = [float(v) for v in " ".join(lines[line_n + 2:line_n + 2 + nlines]).split()]
+    return spin, kpoint, bands
+
+
+def _parse_one_band_after_81(lines, line_n):
```
Comment (yakutovicha) [inline aiida_cp2k/utils/parser.py:299]:
```suggestion
def _parse_bands_cp2k_greater_81(lines, line_n):
```

File: aiida_cp2k/utils/parser.py [hunk 4]
```diff
@@ -282,8 +285,34 @@ def parse_cp2k_output_advanced(fstring):  # pylint: disable=too-many-locals, too
     return result_dict
 
 
-def _parse_bands(lines, n_start):
-    """Parse band structure from cp2k output"""
+def _parse_one_band_before_81(lines, line_n):
```
Comment (yakutovicha) [inline aiida_cp2k/utils/parser.py:288]:
```suggestion
def _parse_bands_cp2k_lower_81(lines, line_n):
```

File: aiida_cp2k/utils/parser.py [hunk 5]
```diff
@@ -292,31 +321,33 @@ def _parse_bands(lines, n_start):
     bands_s1 = []
     bands_s2 = []
     known_kpoints = {}
-    pattern = re.compile(".*?Nr.*?Spin.*?K-Point.*?", re.DOTALL)
+
+    if cp2k_version < 8.1:
+        parse_one_band = _parse_one_band_before_81
```
Comment (yakutovicha) [inline aiida_cp2k/utils/parser.py:326]:
```suggestion
        parse_one_band = _parse_bands_cp2k_lower_81
```

File: aiida_cp2k/utils/parser.py [hunk 6]
```diff
@@ -292,31 +321,33 @@ def _parse_bands(lines, n_start):
...
+        pattern = re.compile(".*?Nr.*?Spin.*?K-Point.*?", re.DOTALL)
+    else:
+        parse_one_band = _parse_one_band_after_81
```
Comment (yakutovicha) [inline aiida_cp2k/utils/parser.py:329]:
```suggestion
        parse_one_band = _parse_bands_cp2k_greater_81
```

**Comment by yakutovicha** [comment]
> Hi @yakutovicha ,
> 
> I tried to implement your suggestions. I'm not sure if I did it the way you intended.
> Could you please have another look at the implementation?

@hdsassnick I decided to continue the development a little bit on my own - sorry for that. I just thought that it would be easier for me to implement what I had in mind directly rather than explain what I actually meant.

> ... e.g. the number of bands is not printed at the head of each k-point in the new cp2k version.

I think that is a pretty minor difference which could be resolved by employing `try...except` - please have a look in the code.

> That is why I have a few doubts that merging the two versions of "_parse_bands" would simplify the code by a large extent.

Let me know what you think of the current implementation.

**Comment by hdsassnick** [comment]
Hi @yakutovicha ,
Thanks so much! This new version looks fantastic.
I just would have a small addition concerning the case when labels are not given (I also overlooked this detail in my first commit, sorry).

File: test/test_parser.py [hunk 7]
```diff
@@ -0,0 +1,41 @@
+# -*- coding: utf-8 -*-
+###############################################################################
+# Copyright (c), The AiiDA-CP2K authors.                                      #
+# SPDX-License-Identifier: MIT                                                #
+# AiiDA-CP2K is hosted on GitHub at https://github.com/aiidateam/aiida-cp2k        #
```
Comment (yakutovicha) [inline test/test_parser.py:5]:
```suggestion
# AiiDA-CP2K is hosted on GitHub at https://github.com/aiidateam/aiida-cp2k  #
```

File: test/test_parser.py [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+                kpoints, labels, bands = _parse_bands(lines, i_line, 5.1)
+        assert (kpoints[4] == [0.2, 0., 0.2]).all()
+        assert labels == [(0, 'GAMMA'), (10, 'X'), (11, 'X'), (21, 'U'), (22, 'K'), (32, 'GAMMA'), (33, 'GAMMA'),
+                          (43, 'L'), (44, 'L'), (54, 'W'), (55, 'W'), (65, 'X')]
+        assert (bands[0] == [-6.84282475, 5.23143741, 5.23143741, 5.23143741, 7.89232311]).all()
+
+
+def test_bands_parser_81():
+    """Test that band structure is parsed correctly."""
+    # Test parsing bands in the the output of CP2K 8.1
```
Comment (yakutovicha) [inline test/test_parser.py:32]:
```suggestion
    """Test parsing bands in the the output of CP2K 8.1"""
```

File: test/test_parser.py [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+import os
+from aiida_cp2k.utils.parser import _parse_bands
+
+THISDIR = os.path.dirname(os.path.realpath(__file__))
+
+
+def test_bands_parser_51():
+    """Test that band structure is parsed correctly."""
+
+    # Test parsing bands in the the output of CP2K 5.1
```
Comment (yakutovicha) [inline test/test_parser.py:18]:
```suggestion
    """Test parsing bands in the output of CP2K 5.1"""

```

**Review by yakutovicha — APPROVED** [review]
alright, I think this is ready to go.

Thanks, @hdsassnick for your contribution and for using the plugin. You are very welcome to create issues if something isn't working correctly or issues/PRs in case you want to suggest further improvements!

**Comment by yakutovicha** [comment]
This will be available in the next minor release of the plugin.

**Pull request merged**